File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
from typing import List
class Solution:
    def splitArray(self,nums: List[int], k: int) -> int:
        
        """
        Splits an array into k non-empty subarrays to minimize the largest subarray sum.

        Args:
        nums: A list of integers.
        k: The number of subarrays to split into.

        Returns:
        The minimized largest sum of the split.
        """

        def can_split(largest_sum: int) -> bool:
            """
            Checks if the array can be split into k or fewer subarrays,
            with no subarray sum exceeding largest_sum.
            """
            subarray_count = 1
            current_sum = 0
            for num in nums:
                if current_sum + num > largest_sum:
                    # Start a new subarray
                    subarray_count += 1
                    current_sum = num
                else:
                    current_sum += num
            return subarray_count <= k

        # The search range for the answer
        # The minimum possible largest sum is the max element in nums.
        # The maximum possible largest sum is the sum of all elements.
        low = max(nums)
        high = sum(nums)
        result = high

        while low <= high:
            mid = (low + high) // 2
            
            # If we can split the array with mid as the largest sum
            if can_split(mid):
                # Try for a smaller largest sum
                result = mid
                high = mid - 1
            else:
                # Need to allow a larger sum
                low = mid + 1
                
        return result
            

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
#include <vector>
#include <numeric>    // For std::accumulate
#include <algorithm>  // For std::max_element

class Solution {
public:
    int splitArray(std::vector<int>& nums, int k) {
        // The helper function checks if we can split the array into 'k' or fewer
        // subarrays, where no subarray sum exceeds the given 'largest_sum'.
        auto canSplit = [&](long long largest_sum) {
            int subarray_count = 1;
            long long current_sum = 0;
            for (int num : nums) {
                if (current_sum + num > largest_sum) {
                    // Start a new subarray if the current one would exceed the limit.
                    subarray_count++;
                    current_sum = num;
                } else {
                    current_sum += num;
                }
            }
            return subarray_count <= k;
        };

        // Set the search range for our binary search.
        // 'low' is the largest single element, as that's the smallest possible answer.
        // 'high' is the total sum of the array, the largest possible answer.
        // We use 'long long' to prevent potential integer overflow with large sums.
        long long low = *std::max_element(nums.begin(), nums.end());
        long long high = std::accumulate(nums.begin(), nums.end(), 0LL);
        
        long long result = high;

        while (low <= high) {
            long long mid = low + (high - low) / 2;
            
            if (canSplit(mid)) {
                // This 'mid' is a potential answer. Let's try for an even smaller sum.
                result = mid;
                high = mid - 1;
            } else {
                // This 'mid' is too small. We need to allow a larger subarray sum.
                low = mid + 1;
            }
        }

        return result;
    }
};